File: examples/csharp/Patterns/Prototype.cs
namespace DesignPatterns.Prototype
{
    public abstract class Shape
    {
        public int X { get; set; }
        public int Y { get; set; }
        public string Color { get; set; } = "white";

        public abstract Shape Clone();
    }

    public class Circle : Shape
    {
        public int Radius { get; set; }

        public override Shape Clone() => new Circle
            { X = X, Y = Y, Color = Color, Radius = Radius };

        public override string ToString() =>
            $"Circle(x={X}, y={Y}, radius={Radius}, color={Color})";
    }

    public class Rectangle : Shape
    {
        public int Width { get; set; }
        public int Height { get; set; }

        public override Shape Clone() => new Rectangle
            { X = X, Y = Y, Color = Color, Width = Width, Height = Height };

        public override string ToString() =>
            $"Rectangle(x={X}, y={Y}, width={Width}, height={Height}, color={Color})";
    }

    public static class Demo
    {
        public static void Run()
        {
            var circle = new Circle { X = 10, Y = 20, Radius = 15, Color = "red" };
            var clonedCircle = (Circle)circle.Clone();
            clonedCircle.Radius = 25;
            Console.WriteLine($"Original:  {circle}");
            Console.WriteLine($"Cloned:    {clonedCircle}");

            var rect = new Rectangle { X = 5, Y = 5, Width = 30, Height = 20, Color = "blue" };
            var clonedRect = (Rectangle)rect.Clone();
            clonedRect.Color = "green";
            Console.WriteLine($"Original:  {rect}");
            Console.WriteLine($"Cloned:    {clonedRect}");
        }
    }
}

File: examples/cpp/prototype.cpp
#include <iostream>
#include <memory>
#include <string>

class Shape {
public:
    virtual ~Shape() = default;
    virtual std::unique_ptr<Shape> clone() const = 0;
    virtual std::string describe() const = 0;
};

class Circle : public Shape {
    int radius_;
    std::string color_;
public:
    Circle(int radius, const std::string& color) : radius_(radius), color_(color) {}
    std::unique_ptr<Shape> clone() const override {
        return std::make_unique<Circle>(*this);
    }
    std::string describe() const override {
        return "Circle(radius=" + std::to_string(radius_) + ", color=" + color_ + ")";
    }
};

class Rectangle : public Shape {
    int width_, height_;
    std::string color_;
public:
    Rectangle(int w, int h, const std::string& color) : width_(w), height_(h), color_(color) {}
    std::unique_ptr<Shape> clone() const override {
        return std::make_unique<Rectangle>(*this);
    }
    std::string describe() const override {
        return "Rectangle(" + std::to_string(width_) + "x" + std::to_string(height_) + ", color=" + color_ + ")";
    }
};

int main() {
    std::cout << "=== Prototype: Shape Cloning ===" << std::endl;

    Circle circle(10, "red");
    std::cout << "Original: " << circle.describe() << std::endl;
    auto clonedCircle = circle.clone();
    std::cout << "Clone:    " << clonedCircle->describe() << std::endl;

    Rectangle rect(20, 30, "blue");
    std::cout << "Original: " << rect.describe() << std::endl;
    auto clonedRect = rect.clone();
    std::cout << "Clone:    " << clonedRect->describe() << std::endl;

    return 0;
}
